### PR TITLE
Updating stimulus_table

### PR DESCRIPTION
**Pull request by yavorska-iryna**

### CODE REVIEW

File: databook/visual-behavior/VBO-ExperimentData.md
Comment (yavorska-iryna) [inline databook/visual-behavior/VBO-ExperimentData.md:1]: What's wrong with resetting index? It makes sense to do since image presentation index makes more sense for the table.

**Review by yavorska-iryna — COMMENTED** [review]
I think reseting index after sub-selection makes table index more informative. But I'm ok with either.

**Comment by matchings** [comment]
@yavorska-iryna Resetting the index only changes the `stimulus_presentations_id` by one value. If you reindex, it starts at 0, if you don't, it starts at 1. I don't see how that would change how informative or meaningful it is, so my thought was to reduce number of lines of code for simplicity. 

If anything, it makes it less meaningful because it is changing the original data in the table, which is generally not recommended. 

If someone really wants the index to start at 0, they can reindex it themselves, but i don't think we should be including extra lines of code for an operation that changes the data that is provided by the SDK / NWB files.

**Comment by JEElsner** [comment]
Fixes #48

**Comment by matchings** [comment]
@JEElsner This issue (#41) is only part of the fix related to #48. This PR is about changing the code such that it will work with version 1.1.0 of the Visual Behavior Ophys dataset. The goal of #48 is to make sure that the NWB files associated with v1.1.0 are registered as a data asset in code ocean. The full fix is to put both of those things together so that the code and the files line up. But thanks for identifying the relationship here :)

File: databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md [hunk 1]
```diff
@@ -412,12 +412,13 @@ We can see that as expected, events trace is much cleaner than dff and it genera
 
 ## Stimulus presentations
 
-The `stimulus_presentations` table contains one entry for each stimulus that was presented during the session, along with important metadata including stimulus `start_time` and `end_time`, as well as `image_name`, and whether the stimulus was an image change (`is_change` = True) or an image omission (`omitted` = True).
+The `stimulus_presentations` table contains one entry for each stimulus that was presented during the session, along with important metadata including stimulus `start_time` and `end_time`, as well as `image_name`, and whether the stimulus was an image change (`is_change` = True) or an image omission (`omitted` = True). `stimulus_block` and `stimulus_block_name` indicate the type of stimulus mice were presented at a given point in a session. To select active change detection behavior, first we need to filter the table for `stimulus_block_name` = `change_detection_behavior` or `stimulus_block` = `1`. Note that different sessions may have different number of stimulus blocks, thus `change_detection_behavior` may be assosiated with either `0` or `1` in `stimulus_block` column.
```
Comment (saskiad) [inline databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md:415]: Can we format this the way we do for other tables? Eg. make a list of the column names and their definitions?

Comment (saskiad) [inline databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md:415]: You probably want to add stimulus_block and stimulus_block_name to the list of columns that a few rows down?

I would also recommend showing what all the stimulus block name and stimulus block values are.

File: databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md [hunk 2]
```diff
@@ -412,12 +412,13 @@ We can see that as expected, events trace is much cleaner than dff and it genera
...
 
 Get stimulus information for this experiment and assign it to a table called `stimulus_table`
 
 ```{code-cell} ipython3
 stimulus_table = ophys_experiment.stimulus_presentations.copy()
+stimulus_table = stimulus_table[stimulus_table.stimulus_block_name=='change_detection_behavior']
 stimulus_table.head(10)
 ```
```
Comment (saskiad) [inline databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md:423]: so this example isn't getting stimulus information for the experiment, it's getting stimulus information for a specific block. What about the other blocks?

**Review by saskiad — CHANGES_REQUESTED** [review]
couple small questions

**Comment by matchings** [comment]
@JEElsner  I believe the essential changes have been made for this PR. Would you be able to test the changes in the built databook? 

Also, I realized that the fixes for this PR didn't entirely solve #86 because of the way the plotting functions in that notebook worked, so I implemented a slightly different solution here. Either one would work though.

File: databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md (hunk 1 above)
Comment (matchings) [reply databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md]: fixed

File: databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md (hunk 2 above)
Comment (matchings) [reply databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md:444]: also fixed (same solution as the comment above - i have it print out the different stim blocks now, with more documentation)

File: databook/physiology/stimuli/visual-behavior/VB-Behavior.md [hunk 3]
```diff
@@ -63,31 +72,52 @@ The Visual Behavior Neuropixels dataset used a distinct stimulus configuration t
 
 The inclusion of 2 shared images between the familiar and novel image sets allows for examination of the effects of stimulus context on learned sensory responses. A key question is - do mice (behaviorally, or neurophysiologically) treat the two shared images the same, regardless of whether they are intermixed with other familiar stimuli versus when they are intermixed with novel stimuli?
 
+### Stimulus Novelty 
+
+There are three different types of sessions with natural images, depending on which images were shown during training and which were shown during recording sessions. Assuming 1-8 represent familiar images (shown during training) and a-h represent novel images (shown for the first time during recordings), a given session can be one of the following types: 
+ 
+1) Familiar only sessions (ex: 1111444443333666666)
```
[the remark this answers is not in the document]
Comment (matchings) [reply databook/physiology/stimuli/visual-behavior/VB-Behavior.md]: i added additional explanatory text for this

**Comment by JEElsner** [comment]
Test build started! I'll update you tomorrow with the results.

File: databook/physiology/ophys/visual-behavior/VB-BehaviorSessionData.md
Comment (JEElsner) [inline databook/physiology/ophys/visual-behavior/VB-BehaviorSessionData.md:1]: Loading the cache produces this warning

```
/opt/envs/allensdk/lib/python3.10/site-packages/allensdk/brain_observatory/behavior/behavior_project_cache/behavior_project_cache.py:135: UpdatedStimulusPresentationTableWarning: 
	As of AllenSDK version 2.16.0, the latest Visual Behavior Ophys data has been significantly updated from previous releases. Specifically the user will need to update all processing of the stimulus_presentations tables. These tables now include multiple stimulus types delineated by the columns `stimulus_block` and `stimulus_block_name`.

The data that was available in previous releases are stored in the block name containing 'change_detection' and can be accessed in the pandas table by using: 
	`stimulus_presentations[stimulus_presentations.stimulus_block_name.str.contains('change_detection')]`
  warnings.warn(
```

File: databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md [hunk 4]
```diff
@@ -433,38 +461,81 @@ dataframe columns:
...
 
-duration
-: (float) duration of an image presentation (flash) in seconds (stop_time - start_time). NaN if omitted
+
+Let's explore the `stimulus_presentations` for the `natural_movie_one` stimulus block. Select the `natural_movie_one` block and inspect the unique values of `image_name`, `movie_repeat`, and `movie_frame_index`.
+
+```{code-cell} ipython3
+movie_stimulus_table = stimulus_table[stimulus_table.stimulus_block_name=='natural_movie_one']
+
+print('Values of `image_name` are:', movie_stimulus_table.unique())
```
Comment (JEElsner) [inline databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md:527]: This code cell fails.

```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/tmp/ipykernel_1993/181777896.py in ?()
      1 movie_stimulus_table = stimulus_table[stimulus_table.stimulus_block_name=='natural_movie_one']
      2 
----> 3 print('Values of `image_name` are:', movie_stimulus_table.unique())
      4 print('Values of `movie_repeat` are:', movie_repeat.unique())
      5 print('Values of `movie_frame_index` are:', movie_frame_index.unique())

/opt/envs/allensdk/lib/python3.10/site-packages/pandas/core/generic.py in ?(self, name)
   5898             and name not in self._accessors
   5899             and self._info_axis._can_hold_identifiers_and_holds_name(name)
   5900         ):
   5901             return self[name]
-> 5902         return object.__getattribute__(self, name)

AttributeError: 'DataFrame' object has no attribute 'unique'
```

File: databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md
Comment (JEElsner) [inline databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md:1]: Loading the cache in this file produces the same warning as `VB-BehaviorSessionData.md`.

File: databook/physiology/ophys/visual-behavior/VB-BehaviorSessionData.md
Comment (JEElsner) [inline databook/physiology/ophys/visual-behavior/VB-BehaviorSessionData.md:1]: This page also creates many warnings that clutter up the page in the "Aggregate data across all behavior sessions for this mouse" section. I don't know if that's something we want to clean up by suppressing warnings temporarily, but it would definitely improve readability.

File: databook/physiology/ophys/visual-behavior/VBO-Tutorial-Compare_trial_types.md
Comment (JEElsner) [inline databook/physiology/ophys/visual-behavior/VBO-Tutorial-Compare_trial_types.md:1]: Again, a warning loading the cache.

File: databook/physiology/ophys/visual-behavior/VBO-Tutorial-Compare_trial_types.md [hunk 5]
```diff
@@ -148,11 +148,16 @@ Lets look at the dataframe of stimulus presentations. This tells us the attribut
 stimulus_presentations = ophys_experiment.stimulus_presentations
 stimulus_presentations.head()
 ```
+To select information about stimuli in change detection behavior task only, we need to filter our table by stimulus block name.
+
+```{code-cell} ipython3
+stimulus_presentaions = stimulus_presentations[stimulus_presentaions.stimulus_block_name=='change_detection_behavior'].copy()
```
Comment (JEElsner) [inline databook/physiology/ophys/visual-behavior/VBO-Tutorial-Compare_trial_types.md:154]: This produces a NameError:
```python
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[15], line 1
----> 1 stimulus_presentaions = stimulus_presentations[stimulus_presentaions.stimulus_block_name=='change_detection_behavior'].copy()

NameError: name 'stimulus_presentaions' is not defined
```

**Review by JEElsner — COMMENTED** [review]
Some warnings, a couple errors

File: databook/physiology/ophys/visual-behavior/VB-BehaviorSessionData.md
Comment (matchings) [reply databook/physiology/ophys/visual-behavior/VB-BehaviorSessionData.md:1]: This warning is expected. But isnt really necessary for students at the course to see, so we should look into suppressing it.

File: databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md (hunk 4 above)
Comment (matchings) [reply databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md]: sorry about that, fixed in recent commit

File: databook/physiology/ophys/visual-behavior/VB-BehaviorSessionData.md
Comment (matchings) [reply databook/physiology/ophys/visual-behavior/VB-BehaviorSessionData.md:1]: Yeah it is rather annoying to see this warning every time a dataset is loaded via the SDK. I am not sure how to suppress the warnings temporarily, but it would definitely make things less cluttered.

File: databook/physiology/ophys/visual-behavior/VBO-Tutorial-Compare_trial_types.md (hunk 5 above)
Comment (matchings) [reply databook/physiology/ophys/visual-behavior/VBO-Tutorial-Compare_trial_types.md]: oops, typo! just fixed in last commit

File: databook/physiology/ophys/visual-behavior/VB-BehaviorSessionData.md
Comment (JEElsner) [reply databook/physiology/ophys/visual-behavior/VB-BehaviorSessionData.md:1]: This page doesn't just have the warning at the beginning, there's also many (47) that are caused by the following cell:

```python
behavior_session_ids = this_mouse_table.index.values
behavior_session_dict = {}
for behavior_session_id in behavior_session_ids:
    behavior_session_dict[behavior_session_id] = cache.get_behavior_session(behavior_session_id)
```

It looks like [this page](https://docs.python.org/3/library/warnings.html#temporarily-suppressing-warnings) and [this SO post](https://stackoverflow.com/questions/14463277/how-to-disable-python-warnings) have info on suppressing warnings.

TLDR;
```python
import warnings

# Put this once at the top of the notebook to suppress all warnings for the whole notebook
warnings.filterwarnings('ignore')

# or, prevent specific lines of code from raising warnings
with warnings.catch_warnings(action='ignore'):
    # pesky lines of code here
    pass
```

To play devil's advocate for the warnings though, there's an argument to be made for keeping (at least some) of them so that what the student sees while reading the Databook is the same as what they get when they implement the code themselves.

File: databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md
Comment (JEElsner) [inline databook/physiology/ophys/visual-behavior/VBO-ExperimentData.md:1]: Still getting an error on line 866 in the "Plot example cell activity with stimuli and behavior" section. If the plan is to merge my PR, we can ignore this.

```python
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[46], line 10
      8 # add the colors for each image to the stimulus presentations table in the dataset
      9 stimulus_presentations = ophys_experiment.stimulus_presentations
---> 10 stimulus_presentations['color'] = ophys_experiment.stimulus_presentations['image_name'].map(lambda image_name: colormap[image_name])

File /opt/envs/allensdk/lib/python3.10/site-packages/pandas/core/series.py:4539, in Series.map(self, arg, na_action)
   4460 def map(
   4461     self,
   4462     arg: Callable | Mapping | Series,
   4463     na_action: Literal["ignore"] | None = None,
   4464 ) -> Series:
   4465     """
   4466     Map values of Series according to an input mapping or function.
   4467 
   (...)
   4537     dtype: object
   4538     """
-> 4539     new_values = self._map_values(arg, na_action=na_action)
   4540     return self._constructor(new_values, index=self.index).__finalize__(
   4541         self, method="map"
   4542     )

File /opt/envs/allensdk/lib/python3.10/site-packages/pandas/core/base.py:890, in IndexOpsMixin._map_values(self, mapper, na_action)
    887         raise ValueError(msg)
    889 # mapper is a function
--> 890 new_values = map_f(values, mapper)
    892 return new_values

File /opt/envs/allensdk/lib/python3.10/site-packages/pandas/_libs/lib.pyx:2924, in pandas._libs.lib.map_infer()

Cell In[46], line 10, in <lambda>(image_name)
      8 # add the colors for each image to the stimulus presentations table in the dataset
      9 stimulus_presentations = ophys_experiment.stimulus_presentations
---> 10 stimulus_presentations['color'] = ophys_experiment.stimulus_presentations['image_name'].map(lambda image_name: colormap[image_name])

KeyError: nan
```

**Review by JEElsner — COMMENTED** [review]
closer!

The typo wasn't fully fixed in one place, but I just finished the job myself because it's faster.

File: databook/physiology/ophys/visual-behavior/VBO-Tutorial-Compare_trial_types.md [hunk 6]
```diff
@@ -254,7 +259,7 @@ def plot_stimuli(trial, ax):
     '''
     plot stimuli as colored bars on specified axis
     '''
-    stimuli = ophys_experiment.stimulus_presentations.copy()
+    stimuli = ophys_experiment.stimulus_presentations.query('stimulus_block_name' == @change_detection_behavior').copy()
```
Comment (JEElsner) [inline databook/physiology/ophys/visual-behavior/VBO-Tutorial-Compare_trial_types.md:262]: This line errors:

```python
  Cell In[22], line 19
    stimuli = ophys_experiment.stimulus_presentations.query('stimulus_block_name' == @change_detection_behavior').copy()
                                                                                                               ^
SyntaxError: unterminated string literal (detected at line 19)
```

**Comment by matchings** [comment]
@JEElsner  I think this can be merged if the errors are resolved on your end (thanks for fixing the last few typos)

We should open a separate ticket to address how to deal with the warnings when loading the cache. Thats likely a change to the SDK.

**Comment by JEElsner** [comment]
@matchings Gotcha. There's still a `KeyError` in `VBO-ExperimentData`, but I think that's it.

**Review by d-w-liu — APPROVED** [review]

**Review by d-w-liu — APPROVED** [review]

**Review by JEElsner — APPROVED** [review]
